src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/math/RobustBrentSolver.java

```diff
@@ -0,0 +1,439 @@
+
+/**
+ * A robust version of the Brent solver that tries to avoid spurious non-bracketing conditions.
+ *
+ * The root is bracketed by combing the solution search interval. If multiple roots are found,
+ * the one that maximizes a provided "merit" function is chosen. If a merit function is not provided,
+ * then root is chosen based on a merit "policy" (i.e. largest root, smallest root), in which case,
+ * it is mandatory.
+ *
+ * @author Mehrtash Babadi &lt;mehrtash@broadinstitute.org&gt;
+ */
+public final class RobustBrentSolver {
+
+    /**
+     * If a merit function is not provided, these are the possible policies
+     */
+    public enum MeritPolicy {
+        SMALLEST_ROOT, LARGEST_ROOT
+    }
+
+    private final double relativeAccuracy, absoluteAccuracy, functionValueAccuracy;
+    private int evaluations = 0;
+
+    public RobustBrentSolver(final double relativeAccuracy, final double absoluteAccuracy,
+                             final double functionValueAccuracy) {
+        this.relativeAccuracy = relativeAccuracy;
+        this.absoluteAccuracy = absoluteAccuracy;
+        this.functionValueAccuracy = functionValueAccuracy;
+    }
+
+    /**
+     * The main solver routine
+     *
+     * @param maxEval maximum function evaluations
+     * @param objFunc the objective function to perform root finding on
+     * @param meritFunc a merit function to choose the best root (if multiple roots are found)
+     * @param min lower bound for the root
+     * @param max upper bound for the root
+     * @param numBisections number of bisections for "combing" the search interval
+     * @param depth the fractal depth of the search grid
+     * @return the root
+     * @throws TooManyEvaluationsException if too many calls are made to the objective function
+     * @throws NoBracketingException if the root could not be ultimately bracketed
+     */
+    public double solve(final int maxEval,
+                        @Nonnull final UnivariateFunction objFunc,
+                        @Nullable final UnivariateFunction meritFunc,
+                        @Nullable final MeritPolicy meritPolicy,
+                        final double min, final double max,
+                        final int numBisections, final int depth)
+            throws TooManyEvaluationsException, NoBracketingException {
+        Utils.validateArg(meritFunc != null || meritPolicy != null, "Either a merit function or a merit policy" +
+                        " must be provided");
+        Utils.nonNull(objFunc, "The objective function must be non-null");
+        ParamUtils.isPositive(maxEval, "Maximum function evaluations must be positive");
+        ParamUtils.isPositiveOrZero(max - min, "The search interval is ill-posed");
+        ParamUtils.isPositiveOrZero(numBisections, "Number of search interval bisections must be non-negative");
+        ParamUtils.isPositiveOrZero(depth, "The search depth must be non-negative");
+
+        evaluations = 0;
```

Comment:
I dropped it -- we actually don't keep track of it. It's a useless termination policy anyway.